pkg/virt-launcher/virtwrap/sriov/hostdev_test.go

```diff
@@ -0,0 +1,252 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package sriov_test
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/sriov"
+
+	v1 "kubevirt.io/client-go/api/v1"
+)
+
+var _ = Describe("SRIOV HostDevice", func() {
+	It("creates no device given no interfaces", func() {
+		vmi := &v1.VirtualMachineInstance{}
+
+		Expect(sriov.CreateHostDevices(vmi)).To(BeEmpty())
+	})
+
+	It("creates no device given no SRIOV interfaces", func() {
+		iface := v1.Interface{}
+		iface.Masquerade = &v1.InterfaceMasquerade{}
+		vmi := &v1.VirtualMachineInstance{}
+		vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{iface}
+
+		Expect(sriov.CreateHostDevices(vmi)).To(BeEmpty())
+	})
+
+	It("fails to create device given no available host PCI", func() {
+		iface := newSRIOVInterface("test")
+		vmi := &v1.VirtualMachineInstance{}
+		vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{iface}
+		_, err := sriov.CreateHostDevices(vmi)
+
+		Expect(err).To(HaveOccurred())
+	})
+
+	It("creates no device given bad host PCI address", func() {
+		net := newNetworkData("net1", "resource1", "0bad0pci0address0")
+
+		iface := newSRIOVInterface(net.Name)
+		vmi := &v1.VirtualMachineInstance{}
+		vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{iface}
+
+		var err error
+		env := []envData{net.ResourceEnv, net.DeviceEnv}
+		withEnvironmentContext(env, func() {
+			_, err = sriov.CreateHostDevices(vmi)
+		})
```

Comment:
I see that we exercise the pool and hostdev here, what do you think about making `CreateHostDevice` do the iteration + pop and store it into a struct and then delegate to `createHostDevices(map[iface]string])` this way we really test hostdev here and we don't have to play with env variables.

Reply:
Looking at this again I think we have to test just the singular function `createHostDevice` since the plural `CreateHostDevices` ist just "glue" code between the pool and the `creatHostDevice` it's not worth testing it.

Reply:
I am not in favor of testing unexposed (private) functions. Such tests depend on the implementation details and therefore break more often when the code changes. A test is expected to fail if the logic changes or the API changes, but not when refactoring is done.

I think that the cost here of stubbing the environment is low and does not worth the extra modulaity (detailed in my prev comment).